### PR TITLE
[full-ci] Refactor proxy to get rid of the hardcoded unprotected lists

### PR DESCRIPTION
**Pull request by C0rby**
## Related Issue

- Fixes https://github.com/owncloud/ocis/issues/4401

## How Has This Been Tested?

- test environment:
- test case 1:
- test case 2:
- ...

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link>

### CODE REVIEW

**Comment by kulmann** [comment]
Gave it a spin, with following results:
- uploads work
- preview requests throw a http 500, with error message that reveals an internal 401:
   ```xml
   <d:error xmlns:d="DAV" xmlns:s="http://sabredav.org/ns">
      <s:exception></s:exception>
      <s:message>  {&#34;id&#34;:&#34;com.owncloud.api.thumbnails&#34;,&#34;code&#34;:500,&#34;detail&#34;:&#34;could not get image from source: could not get the image \&#34;1284d238-aa92-42ce-bdc4-0b0000009157$06d8eaca-7a6d-4798-9b31-c24ab6b26a91/SpaceX/starship super heavy.jpeg\&#34;. Request returned with statuscode 401 &#34;,&#34;status&#34;:&#34;Internal Server Error&#34;}</s:message>
   </d:error>
   ```
- creating a text file failed: `PUT` to `https://localhost:9200/remote.php/dav/spaces/1284d238-aa92-42ce-bdc4-0b0000009157%2406d8eaca-7a6d-4798-9b31-c24ab6b26a91/asdf.txt` -> `401 (Unauthorized)`
- opening a previously uploaded text file failed with basic auth popup and 401 on `GET` for `https://localhost:9200/remote.php/dav/spaces/1284d238-aa92-42ce-bdc4-0b0000009157%2406d8eaca-7a6d-4798-9b31-c24ab6b26a91/SpaceX/clm0008.txt`
- renaming files/folders works
- copying filles/folders failed: `COPY` to  `https://localhost:9200/remote.php/dav/spaces/1284d238-aa92-42ce-bdc4-0b0000009157%2406d8eaca-7a6d-4798-9b31-c24ab6b26a91/SpaceX%20asdf` -> `500 (Internal Server Error)` -> after page reload the file files/folders are there. 🤔

**Comment by ownclouders** [comment]
:boom: Acceptance test [<strong>localApiTests-apiArchiver-ocis</strong>](https://drone.owncloud.com/owncloud/ocis/15035/42/1) failed. Further test are cancelled...

**Review by kulmann — APPROVED** [review]
Retested a lot of different scenarios, all worked well. Only thing I found was on the archiver route for public links (see comment), but also broken on master, so IMO not a blocker for this PR.

File: services/proxy/pkg/middleware/authentication.go [hunk 1]
```diff
@@ -91,7 +49,8 @@ func Authentication(auths []Authenticator, opts ...Option) func(next http.Handle
 
 	return func(next http.Handler) http.Handler {
 		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-			if isOIDCTokenAuth(r) || isUnprotectedPath(r) {
+			ri := router.ContextRoutingInfo(r.Context())
+			if isOIDCTokenAuth(r) || ri.IsRouteUnprotected() {
 				// The authentication for this request is handled by the IdP.
```
Comment (rhafer) [inline services/proxy/pkg/middleware/authentication.go:54]: Nit: This comment seems to be not quite correct anymore.

File: services/proxy/pkg/config/defaults/defaultconfig.go [hunk 2]
```diff
@@ -71,20 +71,24 @@ func DefaultPolicies() []config.Policy {
...
 				},
 				{
-					Endpoint: "/signin/",
-					Backend:  "http://localhost:9130",
+					Endpoint:    "/signin/",
+					Backend:     "http://localhost:9130",
+					Unprotected: true,
 				},
 				{
 					Endpoint: "/archiver",
```
[the remark this answers is not in the document]
Comment (rhafer) [reply services/proxy/pkg/config/defaults/defaultconfig.go:94]: Hm, I guess this should be fixable with an addtional `QueryRoute`. I see if I can address this in another PR.

**Review by rhafer — APPROVED** [review]
Looks good to me apart from a little nit picking about the comment.

**Comment by rhafer** [comment]
Turns out that this causes some unstable behavior for routes that match multiple rules :scream:. E.g. we have:

```
                {                                 
                    Endpoint:    "/app/list",
                    Backend:     "http://localhost:9140",
                    Unprotected: true,
                },
                {
                    Endpoint: "/app/", // /app or /apps? ocdav only handles /apps
                    Backend:  "http://localhost:9140",
                },
```

And now requests on `/app/list/` sometimes go through and sometimes fail. (Not this is not exactly a new bug in the code, it just seems that we didn't have overlapping rules of the same type before. I'll open a new issue for that and try to come up with a quick fix. (We might also temporary revert this PR).

**Pull request merged**